src/librustdoc/clean/types.rs

```diff
@@ -2018,15 +2018,23 @@ crate enum GenericArg {
 #[derive(Clone, PartialEq, Eq, Debug, Hash)]
 crate enum GenericArgs {
     AngleBracketed { args: Vec<GenericArg>, bindings: Vec<TypeBinding> },
-    Parenthesized { inputs: Vec<Type>, output: Option<Type> },
+    Parenthesized { inputs: Vec<Type>, output: Option<Box<Type>> },
 }
 
+// `GenericArgs` is in every `PathSegment`, so its size can significantly
+// affect rustdoc's memory usage.
+rustc_data_structures::static_assert_size!(GenericArgs, 56);
+
 #[derive(Clone, PartialEq, Eq, Debug, Hash)]
 crate struct PathSegment {
     crate name: Symbol,
     crate args: GenericArgs,
 }
 
+// `PathSegment` occurs multiple times in every `Path`, so its size can
+// significantly affect rustdoc's memory usage.
+rustc_data_structures::static_assert_size!(PathSegment, 64);
+
```

Comment:
Hmm, this is a good point - I wonder whether it hurts or helps to box GenericArgs instead of `output`? Are you interested in making another perf run with that change? (no problem if not, this is still a good perf improvement as-is)

Reply:
Huh, I just saw your comment now; I wonder if there was a glitch in GitHub.

The thing is that `GenericArgs` is constructed for every `PathSegment`, so I think rustdoc would still be allocating the same amount of memory as without the `Box` (in fact, a bit more, because of the `usize` for the `Box`). Whereas with the change I made, the `Box` is only constructed if (a) the `GenericArgs` is for `Fn` sugar and (b) the `Fn` sugar has an output type.

Reply:
Actually, writing up that comment just gave me an idea for another way to improve perf :D